base/abstractarray.jl

```diff
@@ -891,22 +891,41 @@ end
 ## from general iterable to any array
 
 function copyto!(dest::AbstractArray, src)
-    destiter = eachindex(dest)
-    y = iterate(destiter)
-    for x in src
-        y === nothing &&
+    if haslength(src)
+        length(dest) < length(src) &&
             throw(ArgumentError("destination has fewer elements than required"))
-        dest[y[1]] = x
-        y = iterate(destiter, y[2])
+        i = Int(firstindex(dest))
```

Comment:
It doesn't seem right to me to switch from `eachindex` to `firstindex` just because `src` has a length.

Also, in the past we have avoided annotating `inbounds` in generic methods like this.

Reply:
Ok, I will fiddle a bit more. At the moment, removing `@inbounds` removes the whole speed advantage of this method. But perhaps there's a smarter way.

Reply:
And, I'm think the PR does not handle length zero correctly right now. It breaks these, which work on master, but don't seem to have tests:
```julia
julia> firstindex(Int[])
1

julia> copyto!(Int[], ())
Int64[]

julia> copyto!(Int[], 1, ())
Int64[]
```

Reply:
I've removed this 2-arg method, as on the cases I was testing, I can't see a way to speed it up without using linear indexing & `@inbounds`.

I've also added tests for these empty cases, and fixed the 3-arg method to pass them.